zerver/views/messages.py

```diff
@@ -1108,10 +1108,17 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
 def delete_message_backend(request: HttpRequest, user_profile: UserProfile,
                            message_id: int=REQ(converter=to_non_negative_int)) -> HttpResponse:
     message, ignored_user_message = access_message(user_profile, message_id)
+
     is_user_allowed_to_delete_message = user_profile.is_realm_admin or \
```

Comment:
I think we should extract all the validation logic here to something like `validate_can_delete_message(user_profile, pub_date)` so that we can have direct tests against it.  In particular, we'll want tests to verify that admins can always delete messages.  If you split out the validate function, it allows you to easily cover that scenario:

~~~ py
if user_profile.is_realm_admin:
   # admins can always delete messages, no matter how old
   return

if message.sender != user_profile:
    # normal users can never delete other folks' messages
    raise JsonableError(...) 

if not user_profile.realm.allow_message_deleting:
   raise JsonableError(_"You don't have permission...")

# ...